pkg/pmem-state/pmme-state.go

```diff
@@ -0,0 +1,147 @@
+	GetAll(dataPtr interface{}, foreach GetAllFunc) error
+}
+
+// fileState Persists the state information into a file.
+// This is is supposed to use by Nodes to persists the state.
+type fileState struct {
+	location string
+}
+
+var _ StateManager = &fileState{}
+
+// NewFileState instantiates the file state manager with given directory
+// location. It ensures the provided directory exists.
+// Returns error, if fails to create the direcotry incase of not pre-existing.
+func NewFileState(loc string) (StateManager, error) {
+	if err := ensureLocation(loc); err != nil {
+		return nil, err
+	}
+
+	return &fileState{
+		location: loc,
+	}, nil
+}
+
+// Create saves the volume metadata to file named <id>.json
+func (fs *fileState) Create(id string, data interface{}) error {
+	file := path.Join(fs.location, id+".json")
+	// Create new file for synchronous writes
+	fp, err := os.OpenFile(file, os.O_WRONLY|os.O_SYNC|os.O_CREATE, 0600)
+	if err != nil {
+		return errors.Wrapf(err, "file-state: failed to create metadata storage file %s", file)
+	}
+
+	if err := json.NewEncoder(fp).Encode(data); err != nil {
+		// cleanup file entry before returning error
+		fp.Close() //nolint: errcheck, gosec
+		if e := os.Remove(file); e != nil {
+			glog.Warningf("file-state: fail to remove file %s: %s", file, e.Error())
+		}
+		return errors.Wrap(err, "file-state: failed to encode metadata")
+	}
+
+	return fp.Close()
+}
+
+// Delete deletes the metadata file saved for given volume id
+func (fs *fileState) Delete(id string) error {
+	file := path.Join(fs.location, id+".json")
+	if err := os.Remove(file); err != nil && err != os.ErrNotExist {
+		return errors.Wrapf(err, "file-state: failed to delete file %s", file)
+	}
+
+	return nil
+}
+
+// Get retrieves metadata for given volume id to pointer location of dataPtr
+func (fs *fileState) Get(id string, dataPtr interface{}) error {
+	return fs.readFileData(path.Join(fs.location, id+".json"), dataPtr)
+}
+
+// GetAll retrieves metadata of all volumes found in fileState.location directory.
+// reads all the .json files in fileState.location direcotry and decodes the filedata
+func (fs *fileState) GetAll(dataPtr interface{}, f GetAllFunc) error {
+	files, err := ioutil.ReadDir(fs.location)
+	if err != nil {
+		return errors.Wrapf(err, "file-state: failed to read metadata from %s", fs.location)
+	}
+	for _, fileInfo := range files {
+		fileName := fileInfo.Name()
+		if !strings.HasSuffix(fileName, ".json") {
+			continue
+		}
+
+		file := path.Join(fs.location, fileName)
+		if err := fs.readFileData(file, dataPtr); err != nil {
+			return err
+		}
+
+		id := fileName[0 : len(fileName)-len(".json")]
+		if !f(id) {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func ensureLocation(dir string) error {
+	if _, err := os.Stat(dir); err != nil {
+		if os.IsNotExist(err) {
+			if err := os.Mkdir(dir, 0750); err != nil { //nolint: vetshadow
+				return err
+			}
+		} else {
+			return err
+		}
+	}
```

Comment:
To catch the error early you could check that the dir is actually a directory.

Reply:
Done